File: server/schemas/strategy.py
# -*- coding: utf-8 -*-
"""层级三·策略拓扑与执行计划的 Pydantic 契约。

- StrategyTopology：/strategies 返回项（含 composition/rhythm/capital_allocation）。
- ExecutionPlanNode / ExecutionPlan：/strategies/{name}/plan 返回的依赖树 DAG。
"""
from typing import Any, Dict, List

from pydantic import BaseModel


class StrategyTopology(BaseModel):
    """策略拓扑（GET /strategies 返回项，扩展自原 name/label/universe）。"""
    name: str
    label: str
    universe: List[str]
    composition: Dict[str, Any]            # {"factors": [...], "datasets": [...]}
    rhythm: str                            # 超短频/日频/周频
    capital_allocation: str


class ExecutionPlanNode(BaseModel):
    """执行计划 DAG 节点（数据→因子→信号→下单 生命周期）。"""
    id: str                                # 节点唯一 id（如 data / factor:MACD / signal / order）
    label: str                             # 中文展示名
    stage: str                             # 阶段：data / factor / signal / order
    detail: str = ""                       # 节点详情（多行，前端换行展示）
    depends_on: List[str] = []             # 依赖节点 id 列表（DAG 边）


class ExecutionPlan(BaseModel):
    """GET /strategies/{name}/plan 响应（依赖树渲染「数据→因子→信号→下单」生命周期）。"""
    strategy: str
    label: str
    rhythm: str
    nodes: List[ExecutionPlanNode]

File: web/src/api/strategy.ts
/**
 * 层级三·策略拓扑 API 封装
 *
 * 对应后端 server/api/v1/strategies.py（扩展）。复用 backtest.ts 的 apiClient。
 *
 * 与 backtest.ts 的 getStrategies 区分：
 * - backtest.getStrategies 返回 StrategyMeta（仅 name/label/universe，回测下拉框用）。
 * - 本模块 getStrategies 返回 StrategyTopology（含 composition/rhythm/capital_allocation，
 *   策略架构师视图消费拓扑白盒信息）。
 */
import { apiClient } from './backtest'

/** 策略拓扑（GET /strategies 扩展返回项） */
export interface StrategyTopology {
  name: string
  label: string
  universe: string[]
  composition: { factors?: string[]; datasets?: string[]; [k: string]: unknown }
  rhythm: string
  capital_allocation: string
}

/** 执行计划 DAG 节点 */
export interface ExecutionPlanNode {
  id: string
  label: string
  stage: 'data' | 'factor' | 'signal' | 'order' | string
  detail: string
  depends_on: string[]
}

/** 执行计划 DAG（GET /strategies/{name}/plan） */
export interface ExecutionPlan {
  strategy: string
  label: string
  rhythm: string
  nodes: ExecutionPlanNode[]
}

/** 列出全部策略拓扑（含 composition/rhythm/capital_allocation） */
export function getStrategies(): Promise<StrategyTopology[]> {
  return apiClient.get('/api/v1/strategies', { timeout: 10000 })
}

/** 获取策略执行计划 DAG */
export function getExecutionPlan(name: string): Promise<ExecutionPlan> {
  return apiClient.get(`/api/v1/strategies/${encodeURIComponent(name)}/plan`, { timeout: 10000 })
}
